### PR TITLE
[compiler]add ccl bufferization and lccl dialect.

### PR DESCRIPTION
**Pull request by linuxlonelyeagle**
Add bufferization of ccl dialect and lcc dialect. see the test for the bufferization part.

### CODE REVIEW

File: compiler/dialects/include/byteir/Dialect/Ccl/IR/CclOps.td [hunk 1]
```diff
@@ -217,8 +215,6 @@ def Ccl_SendOp : Ccl_Op<"send", [
 }
 
 def Ccl_RecvOp : Ccl_Op<"recv", [
-    SameOperandsAndResultElementType,
-    SameOperandsAndResultShape,
```
Comment (qingyunqu) [inline compiler/dialects/include/byteir/Dialect/Ccl/IR/CclOps.td:221]: Don't need to remove these when bufferize to lccl dialect?

File: compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp [hunk 2]
```diff
@@ -125,9 +125,6 @@ AllReduceOp::inferReturnTypes(MLIRContext *, std::optional<Location> location,
                               ValueRange operands, DictionaryAttr,
                               OpaqueProperties, RegionRange,
                               SmallVectorImpl<Type> &inferredReturnTypes) {
-  if (operands.size() != 1)
```
Comment (qingyunqu) [inline compiler/dialects/lib/Dialect/Ccl/IR/CclOps.cpp:128]: same to here.

**Review by qingyunqu — CHANGES_REQUESTED** [review]

File: compiler/include/byteir/Dialect/Lccl/LcclOps.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+using ReplicaGroupsIndicesRef = llvm::ArrayRef<int64_t>;
+
+#include "byteir/Dialect/Lccl/LcclOpsDialect.h.inc"
+
+#define GET_OP_CLASSES
+#include "byteir/Dialect/Lccl/LcclOps.h.inc"
+
+namespace mlir {
+namespace lccl {
+constexpr StringRef getRedOpSumName() { return "sum"; }
```
Comment (qingyunqu) [inline compiler/include/byteir/Dialect/Lccl/LcclOps.h:36]: I think it's more simple to reuse ccl's definition.

File: compiler/include/byteir/Dialect/Lccl/LcclOps.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+#ifndef BYTEIR_DIALECT_LCCL_LCCLOPS_H
+#define BYTEIR_DIALECT_LCCL_LCCLOPS_H
+
+#include "mlir/Bytecode/BytecodeOpInterface.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OpImplementation.h"
+
+using ReplicaGroupsIndices = llvm::SmallVector<int64_t>;
+using ReplicaGroupsIndicesRef = llvm::ArrayRef<int64_t>;
```
Comment (qingyunqu) [inline compiler/include/byteir/Dialect/Lccl/LcclOps.h:27]: also reuse here.

**Review by qingyunqu — CHANGES_REQUESTED** [review]

File: compiler/dialects/include/byteir/Dialect/Ccl/IR/CclOps.td (hunk 1 above)
Comment (linuxlonelyeagle) [reply compiler/dialects/include/byteir/Dialect/Ccl/IR/CclOps.td:221]: The dynamic case is considered here.

File: compiler/test/Dialect/Ccl/ccl_bufferize.mlir [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+// CHECK-SAME:      %[[VAL_0:.*]]: memref<4xf32>) -> memref<4xf32> {
+// CHECK:           %[[VAL_1:.*]] = memref.alloc() : memref<4xf32>
+// CHECK:           "lccl.all_reduce"(%[[VAL_0]], %[[VAL_1]]) <{reduction = "sum", replica_groups = {{\[\[}}0, 1], [2, 3]], synchronous = true}> : (memref<4xf32>, memref<4xf32>) -> ()
+// CHECK:           return %[[VAL_1]] : memref<4xf32>
+// CHECK:         }
+
+// -----
+
+func.func @all_reduce_dynamic(%arg0: tensor<4xf32>, %arg1:tensor<1x4xi64>) -> tensor<4xf32> {
+    %0 = "ccl.all_reduce"(%arg0, %arg1) {reduction = "sum", synchronous=true}: (tensor<4xf32>, tensor<1x4xi64>) -> tensor<4xf32>
```
Comment (linuxlonelyeagle) [inline compiler/test/Dialect/Ccl/ccl_bufferize.mlir:150]: @qingyunqu Here is the case of all_reduce dynamic.

File: compiler/dialects/include/byteir/Dialect/Ccl/IR/CclOps.td (hunk 1 above)
Comment (qingyunqu) [reply compiler/dialects/include/byteir/Dialect/Ccl/IR/CclOps.td:232]: If you remove this, you should add check for `ShapedType` in verify function.

**Comment by linuxlonelyeagle** [comment]
I added type inference for broadcast, all_gather,all_reduce,all_reduce_scatter. Added tighter restrictions in bufferize to further determine if there are indeterminate dimensions in the dynamic case, and if so, just stop running the pass.

**Review by qingyunqu — APPROVED** [review]
LGTM, thanks for the efforts.

**Pull request merged**